org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/codemanipulation/StubUtility2Core.java

```diff
@@ -816,53 +768,133 @@ private static String suggestParameterName(ICompilationUnit unit, IVariableBindi
-					if (findMethodBinding(curr, allMethods) == null) {
-						allMethods.add(curr);
+		List<IMethodBinding> allInterfaceMethods= new ArrayList<>();
+
+		for (int i= 0; i < interfaceMethods.length; i++) {
+			for (IMethodBinding method : interfaceMethods[i]) {
+				if (isDefault(method) || isAbstract(method)) {
+					IMethodBinding previouslyFound= findSubSignatureMethod(method, allMethods);
+					if (previouslyFound == null) {
+						IMethodBinding subSigMethod= findSubSignatureMethod(method, allInterfaceMethods);
+						IMethodBinding superSigMethod= findSuperSignatureMethod(method, allInterfaceMethods);
+
+						if (superSigMethod != null && method.overrides(superSigMethod)) {
+							allInterfaceMethods.remove(subSigMethod);
+							allInterfaceMethods.add(method);
+						} else {
+							if (subSigMethod != null && superSigMethod != null) {
+								if (!subSigMethod.getReturnType().getErasure().isSubTypeCompatible(method.getReturnType().getErasure())) {
+									// keep the most specific one
+									allInterfaceMethods.remove(subSigMethod);
+									allInterfaceMethods.add(method);
+								}
+							} else if (superSigMethod != null) {
+								allInterfaceMethods.remove(superSigMethod);
+								allInterfaceMethods.add(method);
+							} else if (subSigMethod == null) {
+								allInterfaceMethods.add(method);
+							}
+						}
 					}
 				}
 			}
-			superClass= superClass.getSuperclass();
 		}
 
-		for (IMethodBinding curr : allMethods) {
-			int modifiers= curr.getModifiers();
-			if ((Modifier.isAbstract(modifiers) || curr.getDeclaringClass().isInterface()) && (implementAbstractsOfInput || typeBinding != curr.getDeclaringClass())) {
-				// implement all abstract methods
-				toImplement.add(curr);
+		allMethods.addAll(allInterfaceMethods);
+
+		return allMethods;
+	}
+
+	private static IMethodBinding findSubSignatureMethod(IMethodBinding method, Collection<IMethodBinding> methods) {
+		for (IMethodBinding candidate : methods) {
+			if (candidate.getName().equals(method.getName()) && candidate.isSubsignature(method)) {
+				return candidate;
 			}
 		}
+		return null;
+	}
 
-		HashSet<ITypeBinding> visited= new HashSet<>();
-		ITypeBinding curr= typeBinding;
-		while (curr != null) {
-			for (ITypeBinding superInterface : curr.getInterfaces()) {
-				findUnimplementedInterfaceMethods(superInterface, visited, allMethods, typeBinding.getPackage(), toImplement);
+	private static IMethodBinding findSuperSignatureMethod(IMethodBinding method, Collection<IMethodBinding> methods) {
+		for (IMethodBinding candidate : methods) {
+			if (candidate.getName().equals(method.getName()) && method.isSubsignature(candidate)) {
+				return candidate;
 			}
-			curr= curr.getSuperclass();
 		}
+		return null;
+	}
+
+	private static boolean isStatic(IMethodBinding method) {
+		return Modifier.isPrivate(method.getModifiers());
+	}
+
+	private static boolean isPrivate(IMethodBinding method) {
+		return Modifier.isStatic(method.getModifiers());
```

Comment:
Ooops.

Reply:
Thanks Stephan for point this out.  I have created: https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/213
to fix this.